File: apps/frontend/evals/chord-detection/label-map.ts
import { CHORDS } from "../../src/data/chords";

export const SUPPORTED_CHORD_IDS = new Set(CHORDS.map((chord) => chord.id));
export const SUPPORTED_CHORD_ID_LIST = [...SUPPORTED_CHORD_IDS].sort();

const ROOT_ALIASES: Record<string, string> = {
  "B#": "C",
  "C#": "C#",
  Db: "C#",
  "D#": "D#",
  Eb: "D#",
  Fb: "E",
  "E#": "F",
  "F#": "F#",
  Gb: "F#",
  "G#": "G#",
  Ab: "G#",
  "A#": "A#",
  Bb: "A#",
  Cb: "B",
};

export function normalizeChordLabel(rawLabel: string): string | null {
  const raw = rawLabel.trim();
  if (!raw || raw.toUpperCase() === "N" || raw.toLowerCase() === "noise") return null;
  const withoutBass = raw.split("/")[0] ?? raw;
  const [rootPart, qualityPart] = withoutBass.includes(":")
    ? splitHarteLabel(withoutBass)
    : splitCompactLabel(withoutBass);
  const root = normalizeRoot(rootPart);
  if (!root) return null;
  const suffix = normalizeQuality(qualityPart);
  if (suffix == null) return null;
  const chordId = `${root}${suffix}`;
  return SUPPORTED_CHORD_IDS.has(chordId) ? chordId : null;
}

function splitHarteLabel(label: string): [string, string] {
  const [root = "", quality = "maj"] = label.split(":", 2);
  return [root, quality];
}

function splitCompactLabel(label: string): [string, string] {
  const match = /^([A-G](?:#|b)?)(.*)$/.exec(label);
  if (!match) return [label, ""];
  return [match[1] ?? "", match[2] ?? ""];
}

function normalizeRoot(root: string): string | null {
  if (!root) return null;
  const normalized = root[0]?.toUpperCase() + root.slice(1);
  return ROOT_ALIASES[normalized] ?? normalized;
}

function normalizeQuality(quality: string): string | null {
  const q = quality.trim();
  if (q === "" || q === "maj" || q === "major") return "";
  if (q === "m" || q === "min" || q === "minor") return "m";
  if (q === "7" || q === "dom7") return "7";
  if (q === "5" || q === "power") return "5";
  return null;
}

File: apps/backend/app/evals/chord_detection/label_map.py
from __future__ import annotations

from .catalog import CHORDS_BY_ID

ROOT_ALIASES = {
    "B#": "C",
    "C#": "C#",
    "Db": "C#",
    "D#": "D#",
    "Eb": "D#",
    "Fb": "E",
    "E#": "F",
    "F#": "F#",
    "Gb": "F#",
    "G#": "G#",
    "Ab": "G#",
    "A#": "A#",
    "Bb": "A#",
    "Cb": "B",
}


def normalize_chord_label(raw_label: str) -> str | None:
    raw = raw_label.strip()
    if not raw or raw.upper() == "N" or raw.lower() == "noise":
        return None
    without_bass = raw.split("/", 1)[0]
    root_part, quality_part = (
        split_harte_label(without_bass) if ":" in without_bass else split_compact_label(without_bass)
    )
    root = normalize_root(root_part)
    if root is None:
        return None
    suffix = normalize_quality(quality_part)
    if suffix is None:
        return None
    chord_id = f"{root}{suffix}"
    return chord_id if chord_id in CHORDS_BY_ID else None


def split_harte_label(label: str) -> tuple[str, str]:
    parts = label.split(":", 1)
    return parts[0], parts[1] if len(parts) > 1 else "maj"


def split_compact_label(label: str) -> tuple[str, str]:
    if len(label) >= 2 and label[1] in {"#", "b"}:
        return label[:2], label[2:]
    return label[:1], label[1:]


def normalize_root(root: str) -> str | None:
    if not root:
        return None
    normalized = root[:1].upper() + root[1:]
    return ROOT_ALIASES.get(normalized, normalized)


def normalize_quality(quality: str) -> str | None:
    q = quality.strip()
    if q in {"", "maj", "major"}:
        return ""
    if q in {"m", "min", "minor"}:
        return "m"
    if q in {"7", "dom7"}:
        return "7"
    if q in {"5", "power"}:
        return "5"
    return None
